File: lang-portal/backend_go/internal/service/study_activities.go
package service

import (
	"database/sql"
	"fmt"
	"lang-portal/internal/models"
	"time"
)

type StudyActivitiesService struct {
	db *models.DB
}

func NewStudyActivitiesService(db *models.DB) *StudyActivitiesService {
	return &StudyActivitiesService{db: db}
}

func (s *StudyActivitiesService) GetStudyActivity(id int) (map[string]interface{}, error) {
	query := `
		SELECT id, name, COALESCE(thumbnail_url, '') as thumbnail_url, 
		       'Practice your vocabulary with flashcards' as description
		FROM study_activities
		WHERE id = ?
	`
	activity := make(map[string]interface{})
	var (
		activityID int
		name, thumbnailURL, description string
	)
	err := s.db.QueryRow(query, id).Scan(
		&activityID, 
		&name, 
		&thumbnailURL,
		&description,
	)
	if err == sql.ErrNoRows {
		return nil, fmt.Errorf("study activity not found")
	}
	if err != nil {
		return nil, err
	}
	activity["id"] = activityID
	activity["name"] = name
	activity["thumbnail_url"] = thumbnailURL
	activity["description"] = description
	return activity, nil
}

func (s *StudyActivitiesService) GetStudyActivitySessions(activityID, page int) ([]map[string]interface{}, *models.Pagination, error) {
	const itemsPerPage = 100
	offset := (page - 1) * itemsPerPage

	// Query to get total pages and items
	var totalItems int
	err := s.db.QueryRow(`
		SELECT COUNT(*) 
		FROM study_sessions 
		WHERE study_activity_id = ?
	`, activityID).Scan(&totalItems)
	if err != nil {
		return nil, nil, err
	}

	// Calculate total pages
	totalPages := (totalItems + itemsPerPage - 1) / itemsPerPage

	// Query to get sessions
	rows, err := s.db.Query(`
		SELECT 
			ss.id, 
			sa.name as activity_name, 
			g.name as group_name, 
			ss.created_at,
			(SELECT COUNT(*) FROM word_review_items wri WHERE wri.study_session_id = ss.id) as review_items_count
		FROM study_sessions ss
		JOIN study_activities sa ON ss.study_activity_id = sa.id
		JOIN groups g ON ss.group_id = g.id
		WHERE ss.study_activity_id = ?
		ORDER BY ss.created_at DESC
		LIMIT ? OFFSET ?
	`, activityID, itemsPerPage, offset)
	if err != nil {
		return nil, nil, err
	}
	defer rows.Close()

	var sessions []map[string]interface{}
	for rows.Next() {
		session := make(map[string]interface{})
		var (
			id int
			activityName, groupName string
			startTime time.Time
			reviewItemsCount int
		)
		err := rows.Scan(
			&id,
			&activityName,
			&groupName,
			&startTime,
			&reviewItemsCount,
		)
		if err != nil {
			return nil, nil, err
		}
		session["id"] = id
		session["activity_name"] = activityName
		session["group_name"] = groupName
		session["start_time"] = startTime
		session["end_time"] = startTime // Simplified for this example
		session["review_items_count"] = reviewItemsCount
		sessions = append(sessions, session)
	}

	pagination := &models.Pagination{
		CurrentPage:  page,
		TotalPages:   totalPages,
		TotalItems:   totalItems,
		ItemsPerPage: itemsPerPage,
	}

	return sessions, pagination, nil
}

func (s *StudyActivitiesService) CreateStudySession(groupID, activityID int) (*models.StudySession, error) {
	// Verify group and activity exist
	if err := s.verifyGroupAndActivity(groupID, activityID); err != nil {
		return nil, err
	}

	query := `
		INSERT INTO study_sessions (group_id, study_activity_id, created_at)
		VALUES (?, ?, CURRENT_TIMESTAMP)
		RETURNING id, group_id, created_at, study_activity_id
	`

	var session models.StudySession
	err := s.db.QueryRow(query, groupID, activityID).Scan(
		&session.ID,
		&session.GroupID,
		&session.CreatedAt,
		&session.StudyActivityID,
	)
	if err != nil {
		return nil, err
	}

	return &session, nil
}

func (s *StudyActivitiesService) verifyGroupAndActivity(groupID, activityID int) error {
	// Check if group exists
	var groupExists bool
	if err := s.db.QueryRow("SELECT EXISTS(SELECT 1 FROM groups WHERE id = ?)", groupID).Scan(&groupExists); err != nil {
		return err
	}
	if !groupExists {
		return fmt.Errorf("group with ID %d does not exist", groupID)
	}

	// Check if activity exists
	var activityExists bool
	if err := s.db.QueryRow("SELECT EXISTS(SELECT 1 FROM study_activities WHERE id = ?)", activityID).Scan(&activityExists); err != nil {
		return err
	}
	if !activityExists {
		return fmt.Errorf("study activity with ID %d does not exist", activityID)
	}

	return nil
}

File: lang-portal/backend-python/app/handlers/study_activities.py
from flask import Blueprint, jsonify, request
from app.models.models import StudyActivity, StudySession, Group
from app.schemas.schemas import (
    study_activity_schema,
    study_activities_schema,
    study_session_schema,
    study_sessions_schema,
    pagination_schema
)
from app import db

study_activities_bp = Blueprint('study_activities', __name__, url_prefix='/api/study-activities')

@study_activities_bp.route('', methods=['GET'])
def get_study_activities():
    """Get all study activities."""
    activities = StudyActivity.query.all()
    return jsonify(study_activities_schema.dump(activities))

@study_activities_bp.route('/<int:activity_id>', methods=['GET'])
def get_study_activity(activity_id):
    """Get a specific study activity by ID."""
    activity = StudyActivity.query.get_or_404(activity_id)
    return jsonify({
        'id': activity.id,
        'name': activity.name,
        'description': activity.description or '',
        'thumbnail_url': activity.thumbnail_url or ''
    })

@study_activities_bp.route('/<int:activity_id>/study_sessions', methods=['GET'])
def get_study_activity_sessions(activity_id):
    """Get study sessions for a specific activity."""
    page = request.args.get('page', 1, type=int)
    per_page = 100
    
    # Ensure activity exists
    StudyActivity.query.get_or_404(activity_id)
    
    pagination = StudySession.query.filter_by(study_activity_id=activity_id)\
        .order_by(StudySession.created_at.desc())\
        .paginate(page=page, per_page=per_page)
    
    return jsonify({
        'items': study_sessions_schema.dump(pagination.items),
        'pagination': pagination_schema.dump(pagination)
    })

@study_activities_bp.route('', methods=['POST'])
def create_study_activity():
    """Create a new study activity."""
    data = request.get_json()
    
    # Create new study activity
    activity = StudyActivity(
        name=data.get('name'),
        thumbnail_url=data.get('thumbnail_url', ''),
        description=data.get('description', '')
    )
    
    db.session.add(activity)
    db.session.commit()
    
    return jsonify({
        'id': activity.id,
        'name': activity.name,
        'description': activity.description,
        'thumbnail_url': activity.thumbnail_url
    }), 201
